cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationVerifier.java

```diff
@@ -122,18 +127,34 @@ static boolean executeGoalsFor(BalancingConstraint constraint,
                                                     new SystemTime(),
                                                     new MetricRegistry(),
                                                     EasyMock.mock(Executor.class));
-    GoalOptimizer.OptimizerResult optimizerResult = goalOptimizer.optimizations(clusterModel,
-                                                                                goalByPriority,
-                                                                                new OperationProgress());
+    // First optimize with all hard goals if there is any.
+    List<Goal> hardGoals =  goalByPriority.stream().filter(Goal::isHardGoal).collect(Collectors.toList());
+    GoalOptimizer.OptimizerResult optimizerResult1 = hardGoals.isEmpty() ? null
+                                                                         : goalOptimizer.optimizations(clusterModel,
+                                                                                                       hardGoals,
+                                                                                                       new OperationProgress());
+
+    // If provided goal list contains any soft goals, then optimize with all goals.
+    // In this iteration, all the hard goals will not generate any new proposal since they are already satisfied in previous
+    // iteration (otherwise an exception has already been thrown). The optimize result of this iteration will only contains
+    // proposals generated by soft goals.
```

Comment:
The explanation is correct, but this change might introduce potential issues:
1. We no longer test the optimization process with a mix of soft and hard goals.
2. In particular, we no longer test the cases, where soft goals have higher priority compared to hard goals.
-- 
Does it make sense to:
(1) preserve the existing test cases with the original list of goals and
(2) introduce "selected" (i.e. targeted) test cases for 
    2.1. testing each soft goal in a cluster state such as the one exemplified in the corresponding ticket of this PR? -- i.e. a case that the goal needs to move non-immigrant replica(s) to achieve balance. But, for correctness, it should not move.
    2.2. testing each hard goal in a cluster state such that the goal needs to move non-immigrant replica(s) to achieve balance. And, for correctness, it indeed moves replicas.

Reply:
Good point, I am adding a new switch to `OptimizationVerifier` to specify wether optimize with provided goals as a whole or optimize in a two-pass way(first optimize with all hard goals, then optimize with all soft goals).